command line interface app/shoe tutorial (#912)
urbit/docs: tutorials/arvo/cli-tutorial.md

into. Finally in [`%shoe` app walkthrough](#shoe-app-walkthrough) we explore
the functionality of the `%shoe` app and then go through the code line-by-line.

This tutorial can be
considered to be an application equivalent of the [Hoon school
lesson](@/docs/tutorials/hoon/generators.md#ask) on `sole` and `%ask`
generators, which only covers the bare minimum necessary to write generators
that take user input.

## The `shoe` library {#the-shoe-library}

Here we describe how sessions are identified, the specialized `card`s that Gall agents
with the `shoe` library are able to utilize, and the different cores of `/lib/shoe.hoon` and their purpose.

### Session identifiers

An app using the `shoe` library will automatically track sessions by their `sole-id=@ta`.
These are opaque identifiers generated by the connecting client. An app using the `shoe`
library may be connected to by a local or remote ship in order to send commands,
and each of these connections is assigned a unique `@ta` that identifies the
ship and which session on that ship if there are multiple.

### `%shoe` `card`s

Gall agents with the `shoe` library are able to utilize `%shoe` `card`s. These

> additions to the standard set of `cards` have the following shape.

Reviewer: This is super nitpicky, and it's more of a comment than an actual "pls change this" suggestion, but:

```suggestion
additions to the standard set of `cards` have the following shape:
```

I always think `:` is more appropriate, whenever some list or example follows, and you're actively pointing to it. This applies to other such cases in this document, too.

Author: I no good grammar, but I do recall that my habit in this regard was formed by never letting a sentence end without a period being highest priority, and that doing this sometimes resulted in that depending on the structure of the list that followed. I'll look up what the proper style is supposed to be.

The list of possible `sole-effects` can be found in `/sur/sole.hoon`. A few
commonly used ones are as follows.
- `[%txt tape]` is used to display a line of text.
- `[%bel ~]` is used to emit a beep.
- `[%pro sole-prompt]` is used to set the prompt.
- `[%mor (list sole-effect)]` is used to emit multiple effects.

For example, a `sole-effect` that beeps and displays `This is some text.` would
be structured as
```hoon
[%mor [%txt "This is some text."] [%bel ~] ~]
```

## `%shoe` app walkthrough {#shoe-app-walkthrough}

Here we explore the capabilities of the `%shoe` example app and then go through
the code, explaining what each line does. 

### Playing with `%shoe`

First let's test the functionality of `%shoe` so we know what we're getting
into.

Start two fake ships, one named `~zod` and the other can have any name - we will
go with `~nus`. Fake ships run locally are able to see each other, and our
intention is to connect their `%shoe` apps.

On each fake ship start `%shoe` by entering `|start %shoe` into dojo. This will
automatically

> change the prompt to `~zod:shoe>` and `~nus:shoe>`. Enter `demo` from `~zod` and press Enter:

Reviewer: >press Enter

I was gonna say it's "press return", but, you don't need to. Part of the demo is that the command runs instantly upon typing it in. Was this not your experience?

```suggestion
change the prompt to `~zod:shoe>` and `~nus:shoe>`. Type `demo` and watch the following appear:
```

Author: I do have a vivid memory of needing to press Enter, and then a few days later being surprised to find that I no longer needed to. I haven't been sure of what to make of it, whether my memory is faulty, my perception was wrong, my computer was lagging, a bug somewhere, etc. But knowing that not needing to press Enter is the intended behavior is good enough for this edit at least.

Next we implement all of the arms required for a `shoe` agent. Starting with
the standard Gall arms:

```hoon
++  on-init   on-init:def
++  on-save   !>(state)
++  on-load
  |=  old=vase
  ^-  (quip card _this)
  [~ this]
::
++  on-poke   on-poke:def
++  on-watch  on-watch:def
++  on-leave  on-leave:def
++  on-peek   on-peek:def
++  on-agent  on-agent:def
++  on-arvo   on-arvo:def
++  on-fail   on-fail:def
```
These are minimalist Gall app arm implementations using the default behavior found in `def`.

Here begins the implementation of the additional arms required by the
`(shoe:shoe command)` interface.
```hoon
++  command-parser
  |=  sole-id=@ta
  ^+  |~(nail *(like [? command]))
  (cold [& ~] (jest 'demo'))
```
`+command-parser` is of central importance - it is what is used to parse user
input and transform it into `command`s for the app to execute. Writing a proper
command parser requires understanding of the Hoon parsing functions found in the
standard library. How to do so may be found in the parsing documentation (coming
soon). For now, it is sufficient to know that this arm matches the text "demo" and

> produces a `[? command]`-shaped noun in response.

Reviewer: ```suggestion
produces a `[? command]`-shaped noun in response. Note how the `&` signifies that the command will be run as soon as it has been entered, without waiting for the user to press return.
```

Author: Oh this explains why I remember having to press Enter. I was using an older version of the code at the time that did not have this flag (which I wasn't totally sure what it did either until now).

put through the parser (`+command-parser`) and results in a noun of
`command-type` that the underlying application specifies, which shoe then feeds back into the underlying app as an `+on-command` callback.

In addition to the ten arms that all Gall core apps possess, `+shoe` has a few
more specific to making use of the `shoe` library. Thus you will need to wrap the `shoe:shoe` core with the `agent:shoe` core to obain a standard
10-arm Gall agent core. See the [shoe example app
walkthrough](#shoe-example-app-walkthrough) for how to do this.

Under each header we give the type signature for the arm.

#### `+command-parser`

```hoon
  ++  command-parser
    |~  sole-id=@ta
    |~(nail *(like [? command-type]))
```

Input parser for a specific command-line session. If the head of the result is true,
instantly run the command. If it's false, require the user to press return.

#### `+tab-list`

```hoon
  ++  tab-list
    |~  sole-id=@ta
    *(list (option:auto tank))
```

Autocomplete options for the command-line session (to match `+command-parser`).

#### `+on-command`

```hoon
  ++  on-command
    |~  [sole-id=@ta command=command-type]

>     *(quip card _^|(..on-init))

Reviewer: Wondering if we should lie and say `_this` in place of `_^|(..on-init)` instead. It's semantically equivalent, and more likely to be familiar. The current thing suffers from lack of context.

Author: Yeah I have to think for a moment to see why they're the same. `_this` is much better.

```

Called when a valid command is run.

#### `+can-connect`

```hoon
  ++  can-connect
    |~  sole-id=@ta
    *?
```

> Called to determine whether the app may be connected to. For example, you
> may only want the local ship to be able to connect.

Reviewer: ```suggestion
Called to determine whether a session may be opened or connected to. For example, you
may only want the local ship to be able to connect.
```

#### `+on-connect`

```hoon
  ++  on-connect
    |~  sole-id=@ta
    *(quip card _^|(..on-init))
```

> Called when a connection is to be made.

Reviewer: ```suggestion
Called when a session is opened or connected to.
```

#### `+on-disconnect`

```hoon
  ++  on-disconnect
    |~  sole-id=@ta
    *(quip card _^|(..on-init))
```

> Called when a previously made connection is to be disconnected.

Reviewer: ```suggestion
Called when a previously made session gets disconnected from.
```

### `default` core

This core contains the bare minimum implementation of the additional `shoe` arms
beyond the 10 standard Gall app ams. It is used

> analogously to how the `default-agent` core is used for Gall apps.

Reviewer: ```suggestion
analogously to how the `default-agent` core is used for regular Gall apps.
```

### `agent` core

> This is a wrapper core designed to wrap the `shoe` core, which has too many
> arms to be a Gall agent core. This turns it into a standard Gall agent core by
> integrating the additional arms into the standard ones.

Reviewer: ```suggestion
This is a function for wrapping a `shoe` core, which has too many
arms to be a valid Gall agent core. This turns it into a standard Gall agent core by
integrating the additional arms into the standard ones.
```

## The `sole` library {#the-sole-library}

> `shoe` apps create `shoe-effect` cards of the `[%shoe
> ...]` shape, which currently just wrap `sole-effect`s, i.e. instructions for displaying text and producing other effects in the console.

Reviewer: ```suggestion
`shoe` apps may create specialized cards of the `[%shoe
(list @ta) shoe-effect]` shape, where `shoe-effect` currently just wrap `sole-effect`s, i.e. instructions for displaying text and producing other effects in the console.
```

The list of possible `sole-effects` can be found in `/sur/sole.hoon`. A few

> commonly used ones are as follows. `[%txt p=tape]` is used to display a line of text. `[%bel ~]`
> is used to emit a beep. `[%pro sole-prompt]` is used to set the prompt. `[%mor
> p=(list sole-effect)]` is used to emit multiple effects.

Reviewer: ```suggestion
commonly used ones are as follows.
- `[%txt tape]` is used to display a line of text.
- `[%bel ~]` is used to emit a beep.
- `[%pro sole-prompt]` is used to set the prompt.
- `[%mor (list sole-effect)]` is used to emit multiple effects.
```

For example, a `sole-effect` that beeps and displays `This is some text.` would
be structured as
```hoon
[%mor [%txt "This is some text."] [%bel ~] ~]
```

## `%shoe` app walkthrough {#shoe-app-walkthrough}

Here we explore the capabilities of the `%shoe` example app and then go through
the code, explaining what each line does. 

### Playing with `%shoe`

First let's test the functionality of `%shoe` so we know what we're getting
into.

Start two fake ships, one named `~zod` and the other can have any name - we will
go with `~nus`. Fake ships run locally are able to see each other, and our
intention is to connect their `%shoe` apps.

On each fake ship start `%shoe` by entering `|start %shoe` into dojo. You may
then switch to `%shoe` by pressing `Ctrl-X`, possibly multiple times, which will

> change the prompt to `~zod:shoe>` and `~nus:shoe>`. Enter `demo` from `~zod` and press Enter:

Reviewer: Actually I think `|start`ing an app automatically dumps you into its command line, if available? So no ctrl-xing needed. (But you might still want to include is for later necessity.)

`card` is either an ordinary Gall agent `card` or a `%shoe` `card`, which takes
the shape `[%shoe sole-ids=(list @ta) effect=shoe-effect]`. A `%shoe` `card` is
sent to all `sole`s listed in `sole-ids`, imaking them run the `sole-effect`
specified by `effect` (i.e. printing some text).  Here we can
reference `card:shoe` because of `/+  shoe` at the beginning of the app.

```hoon
=|  state-0
=*  state  -
::
```
Add the bunt value of `state-0` to the head of the subject, then give it the
macro `state`. The `-` here is a lark expression referring to the head of the
subject. This allows us to use `state` to refer to the state elsewhere in the
code no matter what version we're using, while also getting direct access to the
contents of `state` (if it had any).

```hoon
%+  verb  |
%-  agent:dbug
^-  agent:gall
%-  (agent:shoe command)
^-  (shoe:shoe command)
```

This sequence of commands tells the compiler that the following core is a Gall
agent core that may possess arms of the sort defined in `verb`, `agent:dbug`,
and `agent:shoe`. These cores tell the compiler what type signature these arms
must have, and ensure that the arms that are not standard Gall arms are put into

> the context so that the resulting core nests within `agent:gall`.

Reviewer: I [still](https://github.com/urbit/docs/pull/912/#discussion_r456804690) feel this paragraph is incorrect and confusing. The paragraph that directly follows is much better. Did you maybe just forget to delete this?

Author: Yup, its gone now.

As always, the casts here are just reminders of what is being produced. So let's
follow what is being called with the `%` runes. We call `(agent:shoe command)`
on what follows (i.e. the rest of the app),
producing a standard Gall agent core. Then we call wrap the Gall agent core with
`agent:dbug`, endowing it with additional arms useful for debugging, and then
wrap again with `verb`.

```hoon
|_  =bowl:gall

> +*  this  .
>     def   ~(. (default-agent this %|) bowl)
>     des   ~(. (default:shoe this command) bowl)
> ::

Reviewer: ```suggestion
+*  this  .
    def   ~(. (default-agent this %|) bowl)
    des   ~(. (default:shoe this command) bowl)
```

```
This is boilerplate Gall agent core code. We set `this` to be a macro for the

> subject, which is the Gall agent core itself. We set `def` to be a macro for
> `default-agent` door, which also initializes it. Then we set `des`  to be a
> macro for the `default:shoe` door, which also initializes it.

Reviewer: ```suggestion
subject, which is the Gall agent core itself. We set `def` to be a macro for
an initialized `default-agent` door. Then we set `des`  to be a
macro for an initialized `default:shoe` door.
```

or, more succinctly,

```suggestion
subject, which is the Gall agent core itself. We set `def` and `des` to be
macros for initialized `default-agent` and `default:shoe` doors respectively.
```

> Next we implements all of the arms requires for a `shoe` agent. Starting with

Reviewer: ```suggestion
Next we implement all of the arms required for a `shoe` agent. Starting with
```

the standard Gall arms:

```hoon
++  on-init   on-init:def
++  on-save   !>(state)
++  on-load
  |=  old=vase
  ^-  (quip card _this)
  [~ this]
::
++  on-poke   on-poke:def
++  on-watch  on-watch:def
++  on-leave  on-leave:def
++  on-peek   on-peek:def
++  on-agent  on-agent:def
++  on-arvo   on-arvo:def
++  on-fail   on-fail:def
```

> These are boilerplate Gall app arms using the minimum implementation found in `def`.

Reviewer: ```suggestion
These are minimalist Gall app arm implementations using the default behavior found in `def`.
```

> Here begins the implementations of the additional arms required by the

Reviewer: ```suggestion
Here begins the implementation of the additional arms required by the
```

`(shoe:shoe command)` interface.
```hoon
++  command-parser
  |=  sole-id=@ta
  ^+  |~(nail *(like [? command]))
  (cold [& ~] (jest 'demo'))
```
`+command-parser` is of central importance - it is what is used to parse user
input and transform it into `command`s for the app to execute. Writing a proper
command parser requires understanding of the Hoon parsing functions found in the
standard library. How to do so may be found in the parsing documentation (coming

> soon). For now, it is sufficient to know that this arm matches `demo` and

Reviewer: ```suggestion
soon). For now, it is sufficient to know that this arm matches the text "demo" and
```

produces a `[? command]`-shaped noun in response.

```hoon
++  tab-list
  |=  sole-id=@ta
  ^-  (list [@t tank])
  :~  ['demo' leaf+"run example command"]
  ==
```
`+tab-list` is pretty much plug-n-play. For each command you want to be tab
completed, add an entry to the `list` begun by `:~` of the form `[%command
leaf+"description"]`. Now whenever the user types a partial command and presses
tab, the console will display the list of commmands that match the partial
command as well as the descriptions given here.

Thus here we have that starting to type `demo` and pressing tab will result in
the following output in the console:
```
demo  run example command
~zod:shoe> demo
```
with the remainder of `demo` now added to the input line.

Next we have `+on-command`, which is called whenever `+command-parser`
recognizes that `demo` has been entered by a user.
```hoon
++  on-command
  |=  [sole-id=@ta =command]
  ^-  (quip card _this)
```
This is a gate that takes in the `sole-id` corresponding to the session and the
`command` noun parsed by `+command-parser` and returns a `list` of `card`s and

> `_this`, which is Gall agent core including its state.

Reviewer: ```suggestion
`_this`, which is our shoe agent core including its state.
```

```hoon
  =-  [[%shoe ~ %sole -]~ this]
```
This creates a cell of a `%shoe` card that triggers a `sole-effect` given by the head of
the subject `-`, then the Gall agent core `this` - i.e. the return result of
this gate. The use of the `=-` rune means that what follows this
expression is actually run first, which puts the desired `sole-effect` into the
head of the subject.
```hoon
  =/  =tape  "{(scow %p src.bowl)} ran the command"
```
We define the `tape` that we want to be printed.
```hoon
  ?.  =(src our):bowl
    [%txt tape]
  [%klr [[`%br ~ `%g] [(crip tape)]~]~]
```
We cannot just produce the `tape` we want printed, - it needs to fit the
`sole-effect` type. This tells us that if the
origin of the command is not our ship to just print it normally with the `%txt`
`sole-effect`. Otherwise we use `%klr`, which prints it stylistically (here it
makes the text green and bold).

> The following allows either `~zod`, the host ship, and its moons to connect to

Reviewer: ```suggestion
The following allows either `~zod`, or the host ship and its moons, to connect to
```

## The `shoe` library {#the-shoe-library}

Here we describe the different cores of `/lib/shoe.hoon` and their purpose.

An app using the `shoe` library will have `sole-ids=(list @ta)` as part of its
state. These are typically called session ids, and they are identifiers not only
for ships, but for multiple sessions on a given ship. An app using the `shoe`
library may be connected to by a local or remote ship in order to send commands,
and each of these connections is assigned a unique `@ta` that identifies the
ship and which session on that ship if there are multiple.

### `shoe` core

An iron (contravariant) door that defines an interface for Gall agents utilized
the `shoe` library. Use this
core whenever you want to receive input from the user and run a command. The input will get
put through the parser (`+command-parser`) and results in a noun of
`command-type` that the underlying application specifies, then the app calls
that command.

In addition to the ten arms that all Gall core apps possess, `+shoe` has a few
more specific to making use of the `shoe` library. Thus you will often find it
convenient to wrap the `shoe` core with the `agent` core to obain a standard

> 10-arm Gall agent core. See the [shoe example app

Author: I think I like it better the way it is, but the feeling is not strong and I don't think its more or less confusing either way.

Input parser for a specific session. If the head of the result is true,
instantly run the command. We give a brief tutorial on parsing in SECTION??

#### `+tab-list`

Autocomplete options for the sessions (to match `+command-parser`).

#### `+on-command`

Called when a valid command is run.

#### `+can-connect`

Called to determine whether the session can be connected to. For example, you
may only want the local ship to be able to connect to the session.

#### `+on-connect`

Called when a connection to the session is made.

#### `+on-disconnect`

Called when a session is disconnected.

### `default` core

This core contains the bare minimum implementation of the additional `shoe` arms
beyond the 10 standard Gall app ams. It is used
analogously to how the `default-agent` core is used for Gall apps.

### `agent` core

This is a wrapper core designed to take in the `shoe` core that has too many
arms to be a Gall agent core, and turns it into a standard Gall agent core by
moving the additional arms into the context. It endows the agent with additonal
arms in its context used for managing `sole` events and for calling `shoe`-specific arms.

## The `sole` library {#the-sole-library}

> In order to display text, `shoe` creates `$shoe-effect`s which for now are just ` [%sole effect=sole-effect]`s which are eventually broken down into Dill

Author: https://github.com/urbit/docs/pull/912/commits/92a63eda7cf8b4891da7987e0533ea2cc06937a7#diff-6d8f36f0b87617684df85577b6a7144dR46

calls.

> From `sur/sole.hoon`:

Author: https://github.com/urbit/docs/pull/912/commits/92a63eda7cf8b4891da7987e0533ea2cc06937a7#diff-6d8f36f0b87617684df85577b6a7144dR156

```
>=
; ~nus is your neighbor
[linked to [p=~nus q=%shoe]]
```
Now `~zod` will have two `%shoe` sessions running - one local one on `~zod` and
one remote one on `~nus`, which you can access by pressing `Ctrl-X` until you see
`~nus:shoe>` from `~zod`'s console. On the other hand, you should not see
`~zod:shoe>` on `~nus`'s side, since you have not connected `~nus` to `~zod`'s
`%shoe` app. When you enter `demo` from `~nus:shoe>` on
`~zod`'s console you will again see `~zod ran the command`, but this time it
should be in the ordinary font used by the console, signifying that the command
is originating from a remote session. Contrast this with entering `demo` from
`~nus:shoe>` in `~nus`'s console, which will display `~nus ran the command` in
bold green text.

Now try to link to `~zod`'s `%shoe` session from `~nus` by switching to the dojo
on `~nus` and entering `|link ~zod %shoe`. You should see
```
>=
[unlinked from [p=~zod q=%shoe]]
```
and if you press `Ctrl-X` you will not get a `~zod:shoe>` prompt. This is
because the example app is set up to always allow `~zod` to connect (as well as
subject moons if the ship happens to be a planet) but not `~nus`, so this

> message means that `~nus` failed to connect to `~zod`'s `%shoe` session.

Author: As far I know we don't document remote linking anywhere (which is why I had to ask you how to do it), so I definitely find it important to include it here. I will try to trim down the text, though.

> message means that `~nus` failed to connect to `~zod`'s `%shoe` session.

Author: I didn't spend too much thinking about it, but I couldn't think of a nice way to remove more than a couple sentences since its mostly focused on the remote linking aspect and the asymmetry of the `+can-connect` arm. Perhaps remote linking could be covered in a separate tutorial that could be referenced, but it would need an example app, and `shoe` shares the bottom with other possible simplest example apps that allow remote linking, so I don't think that would be a gain. Even telling the user to run `|start` seems important here, since the only other place in the docs that command is ever mentioned is in the Egg Timer tutorial in Hoon School, and I can think of a few good reasons to want to have such an important command be mentioned more than a single time. I remember I had to ask someone (maybe you?) to remind me how to start Gall apps when I started working on this tutorial.

> message means that `~nus` failed to connect to `~zod`'s `%shoe` session.

Reviewer: > Even telling the user to run |start seems important here, since the only other place in the docs that command is ever mentioned is in the Egg Timer tutorial in Hoon School

Considering the egg timer isn't trivially discoverable as "hoon apps 101" (imo) this is a docs smell. (;

### `%shoe`'s code

```hoon
::  shoe: example usage of /lib/shoe
::
::    the app supports one command: "demo".
::    running this command renders some text on all sole clients.
::
/+  shoe, verb, dbug, default-agent
```

`/+` is the Ford rune which imports libraries from the `/lib` directory into
the subject.
 * `shoe` is the `shoe` library.
 * `verb` is a library used to print what a Gall agent is doing.
 * `dbug` is a library of debugging tools. Why do we need this?
 * `default-agent` contains a Gall agent core with minimal implementations of
   required Gall arms.

```hoon
|%
+$  state-0  [%0 ~]
+$  command  ~
::
+$  card  card:shoe
--
```
The types used by the app.

`$state-0` stores the state of the app, which is null as there is no state to

> keep track of. It is good practice (required?) to have a type for state anyways

Reviewer: ```suggestion
keep track of. It is good practice to tack on a version number anyways,
```

in case the app is made stateful at a later time, and indeed we will do this

> when we modify this app later in the tutorial.

Reviewer: That's not in this document though.

> `$command` is typically a set of tagged union types that are possible

Reviewer: ```suggestion
`$command` is typically a set of tagged union types that represent the possible
```

commands that can be entered by the user. Since this app only supports one
command, it is unnecessary for it to have any associated data, thus the command
is represented by `~`.

In a non-trivial context, a `$command` is given by `[%name
data]`, where `%name` is the identifier for the type of command and `data` is

> a type or list of types that contain data needed to execute the command. See

Reviewer: This is most common, but not strictly true. You could imagine something like `[chat-room=@t =action]` or whatever.

`app/chat-cli.hoon` for examples of commands, such as `[%say letter:store]` and
`[%delete path]`.

`$card` is either an ordinary Gall agent `card` or a `%shoe` `card`, which takes
the form `[%shoe sole-ids=(list @ta) effect=shoe-effect]`. A `%shoe` `card` is

> sent to all `sole`s listed in `sole-ids`, instructing them to implement the
> action specified by `effect`. Here we can

Reviewer: ```suggestion
sent to all `sole`s listed in `sole-ids`, making them run the sole effect specified by `effect` (ie, printing some text). Here we can
```

reference `card:shoe` because of `/+  shoe` at the beginning of the app.

```hoon
=|  state-0
=*  state  -
::
```

> Pin the bunt value of `state-0` to the head of the subject, then give it the

Reviewer: ```suggestion
Add the bunt value of `state-0` to the head of the subject, then give it the
```

I think?

Author: What is the difference between adding and pinning?

Reviewer: I think I've only seen pinning in the context of adding faces to things? Looking at `=`-rune docs now though, apparently "pin" _is_ just adding something to the subject, so I guess this is fine either way.

macro `state`. The `-` here is a lark expression referring to the head of the
subject. This allows us to use `state` to refer to the state elsewhere in the

> code no matter what version we're using.

Reviewer: ```suggestion
code no matter what version we're using, while also getting direct access to the contents of `state`, if it had any.
```

```hoon
%+  verb  |
%-  agent:dbug
^-  agent:gall
%-  (agent:shoe command)
^-  (shoe:shoe command)
```
This sequence of commands forges a `agent:gall` core that
possess may arms of the sort defined in `verb,` `agent:dbug`, and `agent:shoe`.
This sequence of commands tells the compiler that the following core is a Gall
agent core that may possess arms of the sort defined in `verb`, `agent:dbug`,
and `agent:shoe`. These cores tell the compiler what sort of form these arms
must have, and ensure that the arms that are not standard Gall arms are put into

> the context so that the resulting core nests within `agent:gall`.

Reviewer: There's duplicate text here, and this section is hard to read regardless.

What's happening here is just a sequence of function calls, where we call the "agent wrapper" functions (for lack of a better term) for all those libraries we imported.

The explanation below is much clearer.

Author: https://github.com/urbit/docs/pull/912/commits/cae94e6de188b62721bb40412c37bf52b0aca6fc#diff-6d8f36f0b87617684df85577b6a7144dL298-R295

Author: Hmmm that didn't seem to link directly to where I wanted it to go...

Author: Single lines seem to work though. Apologies for my lack of GH-fu

https://github.com/urbit/docs/pull/912/commits/cae94e6de188b62721bb40412c37bf52b0aca6fc#diff-6d8f36f0b87617684df85577b6a7144dL298

To be more precise, going from bottom to top, we cast what follows to a
`(shoe:shoe command)` app core,
pass that into a gate that transforms it into a `(agent:shoe command)` core, which is then cast
as an `agent:gall` and passed into the `agent:dbug` gate, which
endows the resulting core (stil a valid Gall agent core) with additional arms
useful for debugging purposes. Finally, the `verb` gate (imported with `/+

> verb`) is called, which allows the agent to print what it is doing.

Reviewer: This is a bit misleading, and incorrect on some counts.

The casts are merely type annotations. The way this is written makes it sounds like some transformation happens when we cast _to_ these types, but they're just reminders that we are writing/producing the types specified.

The _call_ to `+agent:shoe` _produces_ a standard gall agent, and from there on everything we do operates on standard gall agents. There are no extra arms, just agents wrapping around agents and adding their own logic.

Author: Oops, missed linking this yesterday: https://github.com/urbit/docs/pull/912/commits/cae94e6de188b62721bb40412c37bf52b0aca6fc#diff-6d8f36f0b87617684df85577b6a7144dL298

```hoon
|_  =bowl:gall
+*  this  .
    def   ~(. (default-agent this %|) bowl)
    des   ~(. (default:shoe this command) bowl)
::
```
This is boilerplate Gall agent core code. We set `def` to be an alias for the part of the
context where the `default-agent` lives, and set `des`  to be an alias for the part of the

> context where `shoe` library commands live.

Reviewer: idk what's meant by "part of the context where x lives" here. Also earlier you said "macro" for `=*`, so I think you should say "macro" here, too. I think that hints at why this is confusing: we're not just pointing to things, we're using them: initializing doors that we want to use frequently.

Author: https://github.com/urbit/docs/pull/912/commits/cae94e6de188b62721bb40412c37bf52b0aca6fc#diff-6d8f36f0b87617684df85577b6a7144dR304

> Next we begin implementing all of the arms 

Reviewer: ```suggestion
Next we implement all of the arms required for a shoe agent. Starting with the standard gall arms:
```

standard library, but we will cover the bare essentials of parsing here.

`+command-parser` is a gate which takes in a `sole-id=@ta` identifying a `sole`
session and produces a particular sort of gate which takes in a `nail` (parse
input) and returns an `edge` (parse output) that we call a parser or `$rule`.
`+command-parser` produces a specific type of `$rule` whose result is of the
type `[? command]`. Let us elaborate.

 `nail` is the remainder of a parse given by `[p=hair q=tape]` where the
`hair=[p=@ud q=@ud]` represent line and column of the current position of the parse and the
`tape` is the parsing input. An `edge` is a given by `[p=hair q=(unit
[p=* q=nail])]`. `p` says what we have parsed up to this point. `q` is the
result of the parse, which is `~` if nothing valid has yet been parsed or a noun
`p.q` that here will be a value of the `$command` type and `q.q` a `nail` that
contains results that may still be unparsed. Here the
`edge` is cast with `*(like [? command])`. `like [? command]`
produces a generic `edge` which returns a result that is a cell of values of the
`?` and `command` types - this is what forces `+command-parser` to produce a

> `$rule` that returns a value of type `[? command]`.

Reviewer: These are not "the bare essential" lol. I think it's fine to handwave this away as "this is the type for a parser that produces a `[? command]`-shaped noun, see this parsing doc (TODO) for deets".

Author: I agree on wiping the parsing section. I should have a PR up for parsing docs next week anyways that I'll be able to link to from here.

https://github.com/urbit/docs/pull/912/commits/cae94e6de188b62721bb40412c37bf52b0aca6fc#diff-6d8f36f0b87617684df85577b6a7144dR339

Anyways, the gate is
```hoon
(cold [& ~] (jest 'demo'))
```

> `cold` is a parser modifier that accepts a constant noun (here `[& ~]`) and a
> `rule` (`(jest 'demo')`) to produce another `rule`. This `rule` is then a parser that
> produces `[& ~]` whenever the `rule` `(jest 'demo')` triggers. `(jest 'demo')`
> triggers whenever the `tape` in the input `nail` exactly
> matches `demo`.

Reviewer: ```suggestion
`+cold` takes a noun and a parser. The result is a parser that's identical to the parser we passed in, except it always produces our noun (`[& ~]`, matching `[? command]`) if parsing succeeds, instead of whatever the parser itself would produce.
```

> The reason we use a `cord` `demo` as the argument for `jest` rather than a
> `tape` is not too important for this discussion, but the short version is that
> `cord`s are used to represent data that is passed around while `tape`s are for
> when you're ready to start doing string manipulation.

Reviewer: ```suggestion
```

I'd just leave this out, tbh. Seems way off-topic.

```hoon
++  tab-list
  |=  sole-id=@ta
  ^-  (list [@t tank])
  :~  ['demo' leaf+"run example command"]
  ==
```
`+tab-list` is pretty much plug-n-play. For each command you want to be tab
completed, add an entry to the `list` begun by `:~` of the form `[%command
leaf+"description"]`. Now whenever the user types a partial command and presses
tab, the console will display the list of commmands that match the partial
command as well as the descriptions given here.

Thus here we have that starting to type `demo` and pressing tab will result in
the following output in the console:
```
demo  run example command
~zod:shoe> demo
```

> with the remainder of `demo` now displayed in the input line.

Reviewer: ```suggestion
with the remainder of `demo` now added to the input line.
```

> Next we have `+on-command`, which is called whenever `+command-parser`
> recognizes that `demo` has be entered by a user on a connected session.

Reviewer: ```suggestion
Next we have `+on-command`, which is called whenever `+command-parser`
recognizes that `demo` has been entered by a user.
```

```hoon
++  on-command
  |=  [sole-id=@ta =command]
  ^-  (quip card _this)
```
This is a gate that takes in the `sole-id` corresponding to the session and the
`command` noun parsed by `+command-parser` and returns a `list` of `card`s and
`_this`, which is Gall agent core including its state.
```hoon
  =-  [[%shoe ~ %sole -]~ this]
```
This creates a cell of a `%shoe` card that triggers a `sole-effect` given by the head of
the subject `-`, then the Gall agent core `this` - i.e. the return result of
this gate. The use of the `=-` rune means that what follows this
expression is actually run first, which puts the desired `sole-effect` into the
head of the subject.
```hoon
  =/  =tape  "{(scow %p src.bowl)} ran the command"
```

> We pin the `tape` that we want to be printed to the head of the subject.

Reviewer: ```suggestion
We define the `tape` that we want to be printed.
```

```hoon
  ?.  =(src our):bowl
    [%txt tape]
  [%klr [[`%br ~ `%g] [(crip tape)]~]~]
```

> We cannot just put a bare `tape` at the head of the subject to be added to the
> card - it needs to be wrapped as a `sole-effect`. This tells us that if the

Reviewer: ```suggestion
We cannot just produce the `tape` we want printed, - it needs to fit the `sole-effect` type.  
This tells us that if the
```

origin of the command is not our ship to just print it normally with the `%txt`
`sole-effect`. Otherwise we use `%klr`, which prints it stylistically (here it
makes the text green and bold).

> The following allows either `~zod` or the current ship or its moons to connect to `%shoe`.

Reviewer: ```suggestion
The following allows `~zod`, the host ship and its moons to connect to this app's command line interface using `|link`.
```

```hoon
++  can-connect
  |=  sole-id=@ta
  ^-  ?
  ?|  =(~zod src.bowl)
      (team:title [our src]:bowl)
  ==
```

> We use the minimal implementations for the final two `shoe` arms.

Reviewer: ```suggestion
We use the minimal implementations for the final two `shoe` arms, since we don't want to do anything special when users connect or disconnect.
```

```hoon
++  on-connect      on-connect:des
++  on-disconnect   on-disconnect:des
--
```

This concludes our review of the code of the `%shoe` app. To continue learning

> how to build your own CLI app, we recommend checking out `/app/chat-cli.hoon`.

Reviewer: Bad rec lol, chat-cli is old and complicated. Better cli apps Soon™.

Author: Is there something else I should suggest there or just leave it be for now?

Reviewer: Probably leave this out for now? I think expecting users to extrapolate from /app/shoe isn't too big an ask, but maybe I'm underestimating the task. chat-cli is just going to be confusing and frustrating, there's way too much unrelated things in there.

Author: I've developed a lot of my own understanding over the past couple of weeks by browsing `chat-cli.hoon`. There is a lot of irrelevant stuff, but I think most users at this point should be able to pick it out. I don't think I would have gotten a better handle trying to extrapolate from `shoe.hoon`. Since there isn't anything else to point the reader at, for now I want to leave it in.
